### PR TITLE
feat(locker): Track the liquidity locked by token address

### LINKED ISSUE #106: feat(locker): Track the liquidity locked by token address
The first MVP is supposed to be launched without a backend (i.e. no indexer).

However, the typical way to check whether a LP token has been locked is to search by its token address. If we do not index the TokenLocked events, we can't know if a token has been locked or not efficiently.

As such, we should implement a tracking mechanism - similar to the one that exists to track LP locked by user - but by tokens.

Steps: 
- Add a storage variable `token_locks` that maps ContractAddresses to a List<u128> (internal IDs)
- Add  `token_locks_length` and `token_locked_at` entrypoints
- refactor `remove_user_lock` into `remove_lock_from_list(self: @ContractState, lock_id: u128, list: List<u128>`

`remove_lock_from_list` will remove the id of a lock from the list of locks supplied as parameter. The implementation doesn't change, instead of doing `let mut list = self.user_locks.read(owner)` __inside__ the function, we will do it before calling the function, and we will pass the correct list instance to `remove_lock_from_list`.

Of course, you will adapt the test accordingly

### PR DESCRIPTION
**Pull request by Akashneelesh**
Resolves #106 

Covered all the description and have done the changes

### CODE REVIEW

File: contracts/src/locker/interface.cairo [hunk 1]
```diff
@@ -192,4 +192,37 @@ trait ITokenLocker<TContractState> {
...
+    ///
+    /// * `u32` - The number of locks associated with the specified token.
+    ///
+    fn token_locks_length(self: @TContractState, token: ContractAddress) -> u32;
+
+    /// Retrieves the ID of a lock associated with a specified token at a given index.
+    ///
+    /// This function returns the lock ID for a specific token based on the provided index.
+    /// It reads the list of lock IDs from the `token_locks` mapping and returns the lock ID at the specified index.
+    /// If the index is out of bounds, the function should handle this appropriately, possibly by returning an error or a special value.
```
Comment (enitrat) [inline contracts/src/locker/interface.cairo:215]:
```suggestion
    /// If the index is out of bounds, the function panics
```

File: contracts/src/locker/interface.cairo [hunk 2]
```diff
@@ -192,4 +192,37 @@ trait ITokenLocker<TContractState> {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `token` - The address of the token contract.
+    /// * `index` - The index of the lock in the token's lock list.
+    ///
+    /// # Returns
+    ///
+    /// * `u128` - The ID of the lock at the given index for the specified token.
+    ///   Returns an error or a special value if the index is out of bounds.
```
Comment (enitrat) [inline contracts/src/locker/interface.cairo:225]:
```suggestion
```

File: contracts/src/locker/token_locker.cairo [hunk 3]
```diff
@@ -184,7 +185,8 @@ mod TokenLocker {
                             unlock_time: 0
                         }
                     );
-                self.remove_user_lock(lock_id, owner);
+                let mut list = self.user_locks.read(owner);
```
Comment (enitrat) [inline contracts/src/locker/token_locker.cairo:188]:
```suggestion
                let mut user_locks = self.user_locks.read(owner);
```

File: contracts/src/locker/token_locker.cairo [hunk 4]
```diff
@@ -184,7 +185,8 @@ mod TokenLocker {
...
+                self.remove_user_lock(lock_id, list);
```
Comment (enitrat) [inline contracts/src/locker/token_locker.cairo:189]: missing removal of the lock from the token_locks list

File: contracts/src/locker/token_locker.cairo [hunk 5]
```diff
@@ -197,7 +199,8 @@ mod TokenLocker {
             let mut token_lock = self.locks.read(lock_id);
 
             // Update user locks
-            self.remove_user_lock(lock_id, token_lock.owner);
+            let mut list = self.user_locks.read(token_lock.owner);
```
Comment (enitrat) [inline contracts/src/locker/token_locker.cairo:202]:
```suggestion
            let mut user_locks = self.user_locks.read(token_lock.owner);
```

Give variable names with semantics

Comment (enitrat) [inline contracts/src/locker/token_locker.cairo:202]:
```suggestion
            let mut user_locks = self.user_locks.read(token_lock.owner);
```

Give variable names with semantics

File: contracts/src/locker/token_locker.cairo [hunk 6]
```diff
@@ -312,8 +325,7 @@ mod TokenLocker {
         /// Internally, this function reads the list of locks of the specified `owner` from the `user_locks` mapping.
         /// It then iterates over the list and replaces the specified `lock_id` with the last element of the list.
         /// The length of the list is then decremented by one, and the last element of the list is set to zero.
-        fn remove_user_lock(self: @ContractState, lock_id: u128, owner: ContractAddress) {
-            let mut list = self.user_locks.read(owner);
+        fn remove_user_lock(self: @ContractState, lock_id: u128, mut list: List<u128>) {
```
Comment (enitrat) [inline contracts/src/locker/token_locker.cairo:328]: The function name and the documentation were not updated

**Review by enitrat — CHANGES_REQUESTED** [review]
Some stuff missing
- Documentation not updated / not accurate
- The token_locks list is never appended to
- `remove_user_lock` function name and documentation not updated to reflect the new behavior
- Once a token is fully unlocked, it's not removed from the token_locks list

File: contracts/src/locker/token_locker.cairo (hunk 6 above)
Comment (mananbordia) [inline contracts/src/locker/token_locker.cairo:328]: I feel arguments list should be more descriptive like: `user_locks`

```suggestion
        fn remove_user_lock(self: @ContractState, lock_id: u128, mut user_locks: List<u128>) {
```

Comment (mananbordia) [reply contracts/src/locker/token_locker.cairo:328]: Also, ownership of list (user_locks) will be given to this function making it unusable in caller function. Is this expected ?

Comment (enitrat) [reply contracts/src/locker/token_locker.cairo:328]: > I feel arguments list should be more descriptive like: user_locks

The purpose of this PR was to refactor this function to remove an element from _any_ list, not just user locks. Here the part to adapt is more the function name and description that the name of the list passed, which can be anything.

Comment (enitrat) [reply contracts/src/locker/token_locker.cairo:328]: > Also, ownership of list (user_locks) will be given to this function making it unusable in caller function. Is this expected ?

I don't think we ever need to use the list after removing items from it - which only happens upon unlocking a token - so I would say this is the right move

**Comment by Akashneelesh** [comment]
> Some stuff missing
> 
> * Documentation not updated / not accurate
> * The token_locks list is never appended to
> * `remove_user_lock` function name and documentation not updated to reflect the new behavior
> * Once a token is fully unlocked, it's not removed from the token_locks list

Thank you @enitrat , I have addressed all these in the new commit.

File: contracts/src/tests/test_token_locker.cairo [hunk 7]
```diff
@@ -101,14 +103,35 @@ mod test_internals {
...
         lock_list = state.user_locks.read(user);
         assert(lock_list.len() == 2, 'should have 2 elements');
         assert(lock_list.array().unwrap_syscall() == array![1_u128, 3_u128], 'should have 1 and 3');
 
         // Check that the last element is no longer accessible
         assert(lock_list.get(2).unwrap_syscall().is_none(), 'prev len should be none');
     }
+
+    #[test]
+    fn test_remove_token_lock() {
```
Comment (enitrat) [inline contracts/src/tests/test_token_locker.cairo:116]: This test is not useful, as it's just a replication of the test above with another list provided. As we're only testing the internal function and not the behavior of a higher-level function, this doesn't add value

File: contracts/src/locker/token_locker.cairo [hunk 8]
```diff
@@ -196,9 +204,18 @@ mod TokenLocker {
...
+            // Update token locks
+            let mut token_locks = self.token_locks.read(token_lock.token);
+            // Removing user lock
+            self.remove_lock_from_list(lock_id, user_locks);
+            // Removing token lock
+            self.remove_lock_from_list(lock_id, token_locks);
+
             let mut new_owner_locks: List<u128> = self.user_locks.read(new_owner);
             new_owner_locks.append(lock_id);
+            let mut new_token_locks : List<u128> =  self.token_locks.read(token_lock.token);
```
Comment (enitrat) [inline contracts/src/locker/token_locker.cairo:217]: ah, ok I get it: this is redundant, you remove the lock to put it back right after in the same list

**Review by enitrat — CHANGES_REQUESTED** [review]

File: contracts/src/locker/token_locker.cairo [hunk 9]
```diff
@@ -196,9 +204,18 @@ mod TokenLocker {
...
 
             // Update user locks
-            self.remove_user_lock(lock_id, token_lock.owner);
+            let mut user_locks = self.user_locks.read(token_lock.owner);
+            // Update token locks
+            let mut token_locks = self.token_locks.read(token_lock.token);
+            // Removing user lock
+            self.remove_lock_from_list(lock_id, user_locks);
+            // Removing token lock
+            self.remove_lock_from_list(lock_id, token_locks);
```
[the remark this answers is not in the document]
Comment (AkashBalasubramani-Router) [reply contracts/src/locker/token_locker.cairo:213]: Got it

[the remark this answers is not in the document]
Comment (Akashneelesh) [reply contracts/src/locker/token_locker.cairo:213]: Got it

File: contracts/src/locker/token_locker.cairo (hunk 8 above)
Comment (Akashneelesh) [reply contracts/src/locker/token_locker.cairo:217]: True true

File: contracts/src/tests/test_token_locker.cairo (hunk 7 above)
Comment (Akashneelesh) [reply contracts/src/tests/test_token_locker.cairo:116]: Yeah sure, got it

**Review by enitrat — APPROVED** [review]

**Pull request merged**